<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Recovery of Spatial Ecotypes from Bulk Gene Expression Data • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recovery of Spatial Ecotypes from Bulk Gene Expression Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Spatial.html">Recovery of Spatial Ecotypes from Spatial Transcriptomics Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Recovery of Spatial Ecotypes from Bulk Gene Expression Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/vignettes/Recovery_Bulk.Rmd" class="external-link"><code>vignettes/Recovery_Bulk.Rmd</code></a></small>
      <div class="d-none name"><code>Recovery_Bulk.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial, we will demonstrate how to infer spatial ecotype
(SE) abundances from bulk RNA-seq data. Specifically, we will infer SE
abundances from RNA-seq data of TCGA melanoma samples. The gene
expression matrix is available at <a href="https://drive.google.com/open?id=14RwnoxhIkGeLp6GXehmAXLvy8Ii2_COU&amp;usp=drive_fs" class="external-link">SKCM_RNASeqV2.geneExp.tsv</a>,
which is a Transcripts Per Million (TPM) matrix obtained from the <a href="https://gdc.cancer.gov/about-data/publications/pancanatlas" class="external-link">PanCanAtlas</a>.</p>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-bulk-expression-of-tcga-melanoma-samples">Loading bulk expression of TCGA melanoma samples<a class="anchor" aria-label="anchor" href="#loading-bulk-expression-of-tcga-melanoma-samples"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"14RwnoxhIkGeLp6GXehmAXLvy8Ii2_COU"</span><span class="op">)</span>, <span class="st">"SKCM_RNASeqV2.geneExp.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># download data from googledrive</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulkdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"SKCM_RNASeqV2.geneExp.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bulkdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bulkdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co">## Set the first column as row names</span></span>
<span><span class="va">bulkdata</span> <span class="op">&lt;-</span> <span class="va">bulkdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">## Drop the first column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bulkdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span> <span class="co">## TPM matrix</span></span></code></pre></div>
<pre><code><span><span class="co">##       TCGA.3N.A9WB.06 TCGA.3N.A9WC.06 TCGA.3N.A9WD.06 TCGA.BF.A1PU.01</span></span>
<span><span class="co">## A1BG         381.0660         195.182        360.8790        176.3990</span></span>
<span><span class="co">## A1CF           0.0000           0.000          0.7092          0.0000</span></span>
<span><span class="co">## A2BP1          0.0000           0.000          6.3830          1.2987</span></span>
<span><span class="co">## A2LD1        250.1980         160.755         97.1986        163.2340</span></span>
<span><span class="co">## A2M         2209.5200      169237.000      18257.9000       6716.4500</span></span>
<span><span class="co">## A2ML1          7.2698           0.000          0.0000          7.7922</span></span>
<span><span class="co">##       TCGA.BF.A1PV.01</span></span>
<span><span class="co">## A1BG         216.8470</span></span>
<span><span class="co">## A1CF           0.0000</span></span>
<span><span class="co">## A2BP1          0.0000</span></span>
<span><span class="co">## A2LD1         60.8727</span></span>
<span><span class="co">## A2M         1740.5800</span></span>
<span><span class="co">## A2ML1          0.5977</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="se-deconvolution">SE deconvolution<a class="anchor" aria-label="anchor" href="#se-deconvolution"></a>
</h2>
<p>The <a href="../reference/DeconvoluteSE.html">DeconvoluteSE</a>
function infers SE abundances in bulk tissues. Users can choose to
either use the default model, which estimates the abundance of
predefined SEs, or apply a custom model to infer the abundance of newly
defined SEs based on bulk gene expression data. Before prediction, the
gene expression data will be log2-transformed (if the max value exceeds
80), followed by normalization to have a mean of 0 and unit variance for
each gene. To ensure accurate univariance normalization, a minimum
sample size of 20 is recommended. For dataset with fewer samples, we
would recommend to use public datasets as reference to estimate the mean
and standard deviation</p>
<p>The default model is trained on pseudobulk gene expression data and
predicts the abundances of nine predefined SEs along with a nonSE group,
which primarily includes cancer cells and cells not associated with any
SEs. The predicted SE and nonSE abundances for each sample sum to 1,
making them comparable across samples.</p>
<p><strong>Note</strong>: 1) SE abundances are directly comparable
across samples, enabling the analysis of SE distribution patterns under
different conditions. 2) Since the number of cell states varies across
SEs, direct abundance comparisons between different SEs or with the
nonSE group may not be meaningful. SEs represent distinct spatial
multicellular communities, each defined by a unique combination of cell
states.</p>
<div class="section level3">
<h3 id="using-default-model">Using default model<a class="anchor" aria-label="anchor" href="#using-default-model"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sefracs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DeconvoluteSE.html">DeconvoluteSE</a></span><span class="op">(</span><span class="va">bulkdata</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sefracs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      nonSE        SE1        SE2        SE3        SE4</span></span>
<span><span class="co">## TCGA.3N.A9WB.06 0.19380234 0.05372667 0.16735600 0.12581663 0.07216666</span></span>
<span><span class="co">## TCGA.3N.A9WC.06 0.05293689 0.13901340 0.08427617 0.08372047 0.04243859</span></span>
<span><span class="co">## TCGA.3N.A9WD.06 0.27200075 0.17114795 0.06288761 0.05641549 0.10166721</span></span>
<span><span class="co">## TCGA.BF.A1PU.01 0.16899510 0.13618971 0.06501306 0.09269615 0.17313296</span></span>
<span><span class="co">## TCGA.BF.A1PV.01 0.12078283 0.06921834 0.05970319 0.16813008 0.11059648</span></span>
<span><span class="co">## TCGA.BF.A1PX.01 0.21398415 0.04019509 0.08906215 0.04499893 0.08561680</span></span>
<span><span class="co">##                        SE5          SE6          SE7        SE8        SE9</span></span>
<span><span class="co">## TCGA.3N.A9WB.06 0.06884572 6.495692e-02 4.610809e-03 0.16471168 0.08400658</span></span>
<span><span class="co">## TCGA.3N.A9WC.06 0.05623098 6.781589e-03 3.162253e-01 0.15964259 0.05873404</span></span>
<span><span class="co">## TCGA.3N.A9WD.06 0.02720647 8.345475e-02 9.239064e-02 0.07760029 0.05522884</span></span>
<span><span class="co">## TCGA.BF.A1PU.01 0.13404194 3.330034e-16 3.330034e-16 0.02543647 0.20449461</span></span>
<span><span class="co">## TCGA.BF.A1PV.01 0.14669963 1.677658e-01 1.260194e-03 0.06961553 0.08622790</span></span>
<span><span class="co">## TCGA.BF.A1PX.01 0.09930790 7.771710e-02 1.239738e-01 0.12590726 0.09923685</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-custom-model">Using custom model<a class="anchor" aria-label="anchor" href="#using-custom-model"></a>
</h3>
<p>After identifying SEs using <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a> or <a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a>,
users can develop a Non-Negative Matrix Factorization (NMF) model to
deconvolve SEs from bulk tissue samples, following the tutorial
below.</p>
<details><summary><strong>Development of SE deconvolution models</strong>
</summary><p>Users can create an NMF model to deconvolve SEs from bulk gene
expression data. The training data can be derived from pseudo-bulk
mixtures created by aggregating single-cell transcriptomics data. In
this tutorial, we will generate pseudo-bulk mixtures using single-cell
RNA-seq data from a melanoma sample, which is also featured in the <a href="Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell
Gene Expression Data</a>.</p>
<p>The raw count for this sample is available at <a href="https://drive.google.com/open?id=17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei&amp;usp=drive_fs" class="external-link">WU2161_counts.tsv</a>
and SE recovery results are available at <a href="https://drive.google.com/open?id=17kFjOHhmWxWAKm-LDqy6wV27o5lvz8L2&amp;usp=drive_fs" class="external-link">WU2161_RecoveredSEs.tsv</a>.</p>
<p><strong>Loading data</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download single-cell gene expression matrix and SE recovery results.</span></span>
<span><span class="co"># The downloads should be finished within 1min.</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei"</span><span class="op">)</span>, <span class="st">"WU2161_counts.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17kFjOHhmWxWAKm-LDqy6wV27o5lvz8L2"</span><span class="op">)</span>, <span class="st">"WU2161_RecoveredSEs.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell gene expression matrix.</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"WU2161_counts.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co">## Set the first column as row names</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">## Drop the first column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load SE recovery results</span></span>
<span><span class="va">SEs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"WU2161_RecoveredSEs.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">SEs</span> <span class="op">=</span> <span class="va">SEs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># extract SE predictions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>: For demonstration purposes, we used 1,337
cells grouped into SEs to create pseudo-bulk mixtures. While this
limited cell number offers a basic example, it may not fully capture the
diverse characteristics of SEs, potentially affecting the robustness of
model training and subsequent SE recovery. In practice, we recommend
using a more comprehensive dataset that accurately reflects the
properties of SEs, ensuring that the training process results in a
reliable recovery model.</p>
<p><strong>Creating pseudobulk data</strong></p>
<p>The <a href="../reference/CreatePseudobulks.html">CreatePseudobulks</a>
function will be used to create pseudo-bulk mixtures from single-cell
(spatial) transcriptomics data with all cells grouped into SEs. It
samples cell fractions from a Gaussian distribution, sets negative
fractions to 0 and re-normalizes fractions to sum to 1 across all SEs.
Based on the resulting fractions, it samples 1,000 cells per pseudo-bulk
mixture with replacement, aggregates their transcriptomes in non-log
linear space, and normalize the resulting mixture to logarithm CPM using
Seurat’s <code>NormalizeData</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">=</span> <span class="fu"><a href="../reference/CreatePseudobulks.html">CreatePseudobulks</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, groups <span class="op">=</span> <span class="va">SEs</span>, n_mixtures <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Mixtures</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span> <span class="co">## Gene expression matrix of pseudobulks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Fracs</span><span class="op">)</span> <span class="co">## SE fractions in pseudobulks</span></span></code></pre></div>
<p><strong>Training</strong></p>
<p>The <a href="../reference/NMFGenerateW.html">NMFGenerateW</a>
function will be used to train an NMF model for SE deconvolution based
on the provided SE fractions and gene expression matrix. Before applying
NMF, each gene’s expression is scaled to have a mean of 0 and unit
variance (recommended). To meet the non-negativity requirement of NMF,
the expression matrix is transformed using the posneg method. This
transformation splits the expression matrix into two matrices: one
containing only positive values and the other containing the absolute
values of the negative values. These two matrices are then concatenated
to form the final training data for the NMF model.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">=</span> <span class="fu"><a href="../reference/NMFGenerateW.html">NMFGenerateW</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Fracs</span><span class="op">)</span>, <span class="va">result</span><span class="op">$</span><span class="va">Mixtures</span><span class="op">)</span></span>
<span><span class="co">## This step should be done within 1 min</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span></span></code></pre></div>
<p>The resulting W matrix can be used to infer SE sbundances from bulk
gene expression profiles by using the <a href="../reference/DeconvoluteSE.html">DeconvoluteSE</a> function.</p>
</details><p><strong>Deconvolution</strong></p>
<p>Then you can use the new model for SE deconvolution by specifying the
<code>W</code> parameter:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sefracs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DeconvoluteSE.html">DeconvoluteSE</a></span><span class="op">(</span><span class="va">bulkdata</span>, W <span class="op">=</span> <span class="va">W</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-se-abundances">Visualization of SE abundances<a class="anchor" aria-label="anchor" href="#visualization-of-se-abundances"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/HeatmapView.html">HeatmapView</a></span><span class="op">(</span><span class="va">sefracs</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.15</span>, <span class="fl">0.3</span><span class="op">)</span>, </span>
<span>            show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, cluster_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Recovery_Bulk_files/figure-html/visualse-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      parallel  stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SpatialEcoTyper_1.0.0 NMF_0.28              Biobase_2.64.0       </span></span>
<span><span class="co">##  [4] BiocGenerics_0.50.0   cluster_2.1.6         rngtools_1.5.2       </span></span>
<span><span class="co">##  [7] registry_0.5-1        dplyr_1.1.4           RANN_2.6.2           </span></span>
<span><span class="co">## [10] Matrix_1.7-0          data.table_1.16.0     Seurat_5.1.0         </span></span>
<span><span class="co">## [13] SeuratObject_5.0.2    sp_2.1-4              googledrive_2.1.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     shape_1.4.6.1          rstudioapi_0.16.0     </span></span>
<span><span class="co">##   [4] jsonlite_1.8.8         magrittr_2.0.3         magick_2.8.5          </span></span>
<span><span class="co">##   [7] spatstat.utils_3.1-0   rmarkdown_2.28         GlobalOptions_0.1.2   </span></span>
<span><span class="co">##  [10] fs_1.6.4               ragg_1.3.2             vctrs_0.6.5           </span></span>
<span><span class="co">##  [13] ROCR_1.0-11            spatstat.explore_3.3-2 htmltools_0.5.8.1     </span></span>
<span><span class="co">##  [16] curl_5.2.2             sass_0.4.9             sctransform_0.4.1     </span></span>
<span><span class="co">##  [19] parallelly_1.38.0      KernSmooth_2.23-24     bslib_0.8.0           </span></span>
<span><span class="co">##  [22] htmlwidgets_1.6.4      desc_1.4.3             ica_1.0-3             </span></span>
<span><span class="co">##  [25] plyr_1.8.9             plotly_4.10.4          zoo_1.8-12            </span></span>
<span><span class="co">##  [28] cachem_1.1.0           igraph_2.0.3           iterators_1.0.14      </span></span>
<span><span class="co">##  [31] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       </span></span>
<span><span class="co">##  [34] R6_2.5.1               fastmap_1.2.0          clue_0.3-65           </span></span>
<span><span class="co">##  [37] fitdistrplus_1.2-1     future_1.34.0          shiny_1.9.1           </span></span>
<span><span class="co">##  [40] digest_0.6.37          colorspace_2.1-1       S4Vectors_0.42.1      </span></span>
<span><span class="co">##  [43] patchwork_1.2.0        tensor_1.5             RSpectra_0.16-2       </span></span>
<span><span class="co">##  [46] irlba_2.3.5.1          textshaping_0.4.0      progressr_0.14.0      </span></span>
<span><span class="co">##  [49] fansi_1.0.6            spatstat.sparse_3.1-0  httr_1.4.7            </span></span>
<span><span class="co">##  [52] polyclip_1.10-7        abind_1.4-5            compiler_4.4.1        </span></span>
<span><span class="co">##  [55] gargle_1.5.2           doParallel_1.0.17      fastDummies_1.7.4     </span></span>
<span><span class="co">##  [58] highr_0.11             MASS_7.3-60.2          rjson_0.2.22          </span></span>
<span><span class="co">##  [61] tools_4.4.1            lmtest_0.9-40          httpuv_1.6.15         </span></span>
<span><span class="co">##  [64] future.apply_1.11.2    goftest_1.2-3          glue_1.7.0            </span></span>
<span><span class="co">##  [67] nlme_3.1-164           promises_1.3.0         gridBase_0.4-7        </span></span>
<span><span class="co">##  [70] Rtsne_0.17             reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">##  [73] gtable_0.3.5           spatstat.data_3.1-2    tidyr_1.3.1           </span></span>
<span><span class="co">##  [76] utf8_1.2.4             spatstat.geom_3.3-2    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">##  [79] foreach_1.5.2          ggrepel_0.9.6          pillar_1.9.0          </span></span>
<span><span class="co">##  [82] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">##  [85] later_1.3.2            circlize_0.4.16        splines_4.4.1         </span></span>
<span><span class="co">##  [88] lattice_0.22-6         survival_3.6-4         deldir_2.0-4          </span></span>
<span><span class="co">##  [91] tidyselect_1.2.1       ComplexHeatmap_2.20.0  miniUI_0.1.1.1        </span></span>
<span><span class="co">##  [94] pbapply_1.7-2          knitr_1.48             gridExtra_2.3         </span></span>
<span><span class="co">##  [97] IRanges_2.38.1         scattermore_1.2        stats4_4.4.1          </span></span>
<span><span class="co">## [100] xfun_0.47              matrixStats_1.4.1      stringi_1.8.4         </span></span>
<span><span class="co">## [103] lazyeval_0.2.2         yaml_2.3.10            evaluate_0.24.0       </span></span>
<span><span class="co">## [106] codetools_0.2-20       tibble_3.2.1           BiocManager_1.30.25   </span></span>
<span><span class="co">## [109] cli_3.6.3              uwot_0.2.2             xtable_1.8-4          </span></span>
<span><span class="co">## [112] reticulate_1.39.0      systemfonts_1.1.0      munsell_0.5.1         </span></span>
<span><span class="co">## [115] jquerylib_0.1.4        Rcpp_1.0.13            globals_0.16.3        </span></span>
<span><span class="co">## [118] spatstat.random_3.3-1  png_0.1-8              spatstat.univar_3.0-1 </span></span>
<span><span class="co">## [121] pkgdown_2.1.0          ggplot2_3.5.1          dotCall64_1.1-1       </span></span>
<span><span class="co">## [124] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          </span></span>
<span><span class="co">## [127] ggridges_0.5.6         crayon_1.5.3           leiden_0.4.3.1        </span></span>
<span><span class="co">## [130] purrr_1.0.2            GetoptLong_1.0.5       rlang_1.1.4           </span></span>
<span><span class="co">## [133] cowplot_1.1.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
